File: fetch_questions.js
const fs = require('fs');
const path = require('path');
const translate = require('google-translate-api-x');

// HTML entity decoder
const decodeHTML = (str) => {
    return str.replace(/&quot;/g, '"')
              .replace(/&#039;/g, "'")
              .replace(/&amp;/g, '&')
              .replace(/&lt;/g, '<')
              .replace(/&gt;/g, '>')
              .replace(/&ntilde;/g, 'ñ')
              .replace(/&eacute;/g, 'é')
              .replace(/&aacute;/g, 'á')
              .replace(/&iacute;/g, 'í')
              .replace(/&oacute;/g, 'ó')
              .replace(/&uacute;/g, 'ú')
              .replace(/&rsquo;/g, "'");
};

async function fetchQuestions() {
    const TOTAL_NEEDED = 1800;
    const BATCH_SIZE = 50;
    const calls = Math.ceil(TOTAL_NEEDED / BATCH_SIZE);
    let allQuestions = [];

    console.log("Fetching session token...");
    const sessionRes = await fetch("https://opentdb.com/api_token.php?command=request");
    const sessionData = await sessionRes.json();
    let token = sessionData.token;

    for (let i = 0; i < calls; i++) {
        console.log(`Fetching batch ${i+1}/${calls}...`);
        
        try {
            const res = await fetch(`https://opentdb.com/api.php?amount=${BATCH_SIZE}&type=multiple&token=${token}`);
            const data = await res.json();
            
            if (data.response_code !== 0) {
                console.log(`API Error code: ${data.response_code}`);
                if (data.response_code === 4) {
                    console.log("Token Empty. Need reset.");
                    await fetch(`https://opentdb.com/api_token.php?command=reset&token=${token}`);
                    await new Promise(r => setTimeout(r, 2000));
                    continue;
                }
                break;
            }

            const results = data.results;
            
            // Extract all strings to translate
            let stringsToTranslate = [];
            for (let item of results) {
                stringsToTranslate.push(decodeHTML(item.question));
                stringsToTranslate.push(decodeHTML(item.correct_answer));
                for (let inc of item.incorrect_answers) {
                    stringsToTranslate.push(decodeHTML(inc));
                }
            }

            console.log(`Translating ${stringsToTranslate.length} strings...`);
            
            // Translate the array in chunks to avoid max length errors
            let translated = [];
            for (let j = 0; j < stringsToTranslate.length; j += 50) {
                const chunk = stringsToTranslate.slice(j, j + 50);
                try {
                    const res = await translate(chunk, {to: 'is'});
                    translated.push(...res.map(r => r.text));
                } catch (e) {
                    console.log(`Translation error on chunk: ${e.message}`);
                    // Fallback: keep english
                    translated.push(...chunk);
                }
                await new Promise(r => setTimeout(r, 500)); // Slight delay
            }

            // Reconstruct
            let idx = 0;
            for (let item of results) {
                const qText = translated[idx];
                const correct = translated[idx+1];
                const inc1 = translated[idx+2];
                const inc2 = translated[idx+3];
                const inc3 = translated[idx+4];
                idx += 5;

                let options = [correct, inc1, inc2, inc3];
                
                // Shuffle options
                for (let k = options.length - 1; k > 0; k--) {
                    const l = Math.floor(Math.random() * (k + 1));
                    [options[k], options[l]] = [options[l], options[k]];
                }
                
                const correctIndex = options.indexOf(correct);

                allQuestions.push({
                    question: qText.replace(/"/g, '\\"').replace(/\n/g, ' '),
                    options: options.map(o => o.replace(/"/g, '\\"').replace(/\n/g, ' ')),
                    correctIndex
                });
            }

            console.log(`Total questions processed: ${allQuestions.length}`);
            await new Promise(r => setTimeout(r, 2000)); // Respect API limits

        } catch (e) {
            console.log(`Error in loop: ${e.message}`);
            break;
        }
    }

    console.log("Formatting for TypeScript...");
    let tsStr = "";
    for (let q of allQuestions) {
        tsStr += `    { question: "${q.question}", options: ["${q.options[0]}", "${q.options[1]}", "${q.options[2]}", "${q.options[3]}"], correctIndex: ${q.correctIndex} },\n`;
    }

    // Append to file
    const targetFile = path.join(__dirname, 'dulur_global/src/lib/games/kviss/content.ts');
    const fileStr = fs.readFileSync(targetFile, 'utf8');
    
    const targetIndex = fileStr.indexOf('export const KVISS_QUESTIONS_EN');
    let beforeEnd = fileStr.lastIndexOf('];', targetIndex);
    
    if (beforeEnd === -1 || targetIndex === -1) {
        console.log("Could not find the end of KVISS_QUESTIONS array.");
        return;
    }

    const newFileContent = fileStr.slice(0, beforeEnd) + tsStr + fileStr.slice(beforeEnd);
    fs.writeFileSync(targetFile, newFileContent);

    console.log("SUCCESSFULLY APPENDED 1800 QUESTIONS!");
}

fetchQuestions();

File: fetch_questions.py
import requests
import json
import time
import html
import random
from deep_translator import GoogleTranslator

def fetch_and_translate(total_needed=1750):
    all_questions = []
    translator = GoogleTranslator(source='en', target='is')
    
    # OpenTDB session token prevents duplicates
    session_res = requests.get("https://opentdb.com/api_token.php?command=request")
    session_token = session_res.json().get("token")
    
    print(f"Session token: {session_token}")
    
    calls = total_needed // 50
    
    for i in range(calls):
        print(f"Fetching batch {i+1}/{calls}...")
        try:
            res = requests.get(f"https://opentdb.com/api.php?amount=50&type=multiple&token={session_token}")
            data = res.json()
            
            if data['response_code'] != 0:
                print(f"API Error code: {data['response_code']}")
                if data['response_code'] == 4:
                    print("Token Empty. Resetting...")
                    requests.get(f"https://opentdb.com/api_token.php?command=reset&token={session_token}")
                    time.sleep(2)
                    continue
                break
                
            results = data['results']
            
            # Prepare strings to translate
            strings_to_translate = []
            for item in results:
                strings_to_translate.append(html.unescape(item['question']))
                strings_to_translate.append(html.unescape(item['correct_answer']))
                for inc in item['incorrect_answers']:
                    strings_to_translate.append(html.unescape(inc))
            
            print("Translating...")
            # Translate in chunks of 50 to avoid payload too large errors
            translated = []
            chunk_size = 50
            for j in range(0, len(strings_to_translate), chunk_size):
                chunk = strings_to_translate[j:j+chunk_size]
                try:
                    trans_chunk = translator.translate_batch(chunk)
                    translated.extend(trans_chunk)
                except Exception as e:
                    print(f"Translation error: {e}")
                    # Fallback to individual
                    for s in chunk:
                        try:
                            translated.append(translator.translate(s))
                        except:
                            translated.append(s) # keep english if failed
            
            # Reconstruct questions
            idx = 0
            for item in results:
                q_text = translated[idx]
                correct = translated[idx+1]
                inc1 = translated[idx+2]
                inc2 = translated[idx+3]
                inc3 = translated[idx+4]
                idx += 5
                
                options = [correct, inc1, inc2, inc3]
                random.shuffle(options)
                correct_idx = options.index(correct)
                
                all_questions.append({
                    "question": q_text.replace('"', '\\"'),
                    "options": [o.replace('"', '\\"') for o in options],
                    "correctIndex": correct_idx
                })
            
            print(f"Total so far: {len(all_questions)}")
            # Save intermediate progress
            with open("kviss_temp.json", "w", encoding="utf-8") as f:
                json.dump(all_questions, f, ensure_ascii=False, indent=2)
                
        except Exception as e:
            print(f"Error in loop: {e}")
            
        time.sleep(3) # Respect API rate limits

    # Format for TS file insertion
    ts_str = ""
    for q in all_questions:
        ts_str += f'    {{ question: "{q["question"]}", options: ["{q["options"][0]}", "{q["options"][1]}", "{q["options"][2]}", "{q["options"][3]}"], correctIndex: {q["correctIndex"]} }},\n'
    
    with open("kviss_new_batch.txt", "w", encoding="utf-8") as f:
        f.write(ts_str)
        
    print("DONE! Wrote to kviss_new_batch.txt")

if __name__ == "__main__":
    fetch_and_translate(1750)
